<h1>Capstone Project</h1>

<p>This is a notebook for my Coursera Applied Data Science Capstone Project</p>

<h2>Objective</h2>

<p>Gathering valuable information and insights on car accidents in Seattle. Then using some of this information to create a machine learning model that is capable of predicting the severity of traffic accidents based on certain attributes.</p>

<font size="+1"><b>Importing Libraries</b></font>

In [1]:
import pandas as pd
import numpy as np

<h2>Load the Data</h2>

In [19]:
df = pd.read_csv(r'Data-Collisions.csv')

C:\Users\fredr\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (33) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<h2>Data Exploration and Cleaning</h2>

In [20]:
pd.set_option("max_columns", 50)
df.head()

SEVERITYCODE           X          Y  OBJECTID  INCKEY  COLDETKEY REPORTNO  \
0             2 -122.323148  47.703140         1    1307       1307  3502005   
1             1 -122.347294  47.647172         2   52200      52200  2607959   
2             1 -122.334540  47.607871         3   26700      26700  1482393   
3             1 -122.334803  47.604803         4    1144       1144  3503937   
4             2 -122.306426  47.545739         5   17700      17700  1807429   

    STATUS      ADDRTYPE   INTKEY  \
0  Matched  Intersection  37475.0   
1  Matched         Block      NaN   
2  Matched         Block      NaN   
3  Matched         Block      NaN   
4  Matched  Intersection  34387.0   

                                      LOCATION EXCEPTRSNCODE EXCEPTRSNDESC  \
0                   5TH AVE NE AND NE 103RD ST                         NaN   
1   AURORA BR BETWEEN RAYE ST AND BRIDGE WAY N           NaN           NaN   
2  4TH AVE BETWEEN SENECA ST AND UNIVERSITY ST           NaN           NaN   
3     2ND AVE BETWEEN MARION ST AND MADISON ST                         NaN   
4              SWIFT AVE S AND SWIFT AV OFF RP           NaN           NaN   

   SEVERITYCODE.1                    SEVERITYDESC COLLISIONTYPE  PERSONCOUNT  \
0               2                Injury Collision        Angles            2   
1               1  Property Damage Only Collision     Sideswipe            2   
2               1  Property Damage Only Collision    Parked Car            4   
3               1  Property Damage Only Collision         Other            3   
4               2                Injury Collision        Angles            2   

   PEDCOUNT  PEDCYLCOUNT  VEHCOUNT                 INCDATE  \
0         0            0         2  2013/03/27 00:00:00+00   
1         0            0         2  2006/12/20 00:00:00+00   
2         0            0         3  2004/11/18 00:00:00+00   
3         0            0         3  2013/03/29 00:00:00+00   
4         0            0         2  2004/01/28 00:00:00+00   

                  INCDTTM                             JUNCTIONTYPE  \
0    3/27/2013 2:54:00 PM   At Intersection (intersection related)   
1   12/20/2006 6:55:00 PM  Mid-Block (not related to intersection)   
2  11/18/2004 10:20:00 AM  Mid-Block (not related to intersection)   
3    3/29/2013 9:26:00 AM  Mid-Block (not related to intersection)   
4    1/28/2004 8:04:00 AM   At Intersection (intersection related)   

   SDOT_COLCODE                                       SDOT_COLDESC  \
0            11  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   
1            16  MOTOR VEHICLE STRUCK MOTOR VEHICLE, LEFT SIDE ...   
2            14       MOTOR VEHICLE STRUCK MOTOR VEHICLE, REAR END   
3            11  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   
4            11  MOTOR VEHICLE STRUCK MOTOR VEHICLE, FRONT END ...   

  INATTENTIONIND UNDERINFL   WEATHER ROADCOND                LIGHTCOND  \
0            NaN         N  Overcast      Wet                 Daylight   
1            NaN         0   Raining      Wet  Dark - Street Lights On   
2            NaN         0  Overcast      Dry                 Daylight   
3            NaN         N     Clear      Dry                 Daylight   
4            NaN         0   Raining      Wet                 Daylight   

  PEDROWNOTGRNT  SDOTCOLNUM SPEEDING ST_COLCODE  \
0           NaN         NaN      NaN         10   
1           NaN   6354039.0      NaN         11   
2           NaN   4323031.0      NaN         32   
3           NaN         NaN      NaN         23   
4           NaN   4028032.0      NaN         10   

                                          ST_COLDESC  SEGLANEKEY  \
0                                  Entering at angle           0   
1  From same direction - both going straight - bo...           0   
2                             One parked--one moving           0   
3                   From same direction - all others           0   
4                                  Enterin

 <font size="+1"><b>Rows and Columns</b></font>

In [21]:
df.shape

(194673, 38)

<p><b>The data set contains 38 columns (severitycode is duplicated) and 194673 rows.</b></p>

 <font size="+1"><b>Timeframe</b></font>

In [24]:
df.sort_values(by='INCDTTM')[['INCDTTM']].head()

INCDTTM
7996   1/1/2004
10714  1/1/2004
5083   1/1/2004
3291   1/1/2004
5940   1/1/2004

<b>From 1/1/2004</b>

In [25]:
df.sort_values(by='INCDTTM', ascending=False)[['INCDTTM']].head()

INCDTTM
192339  9/9/2019 9:27:00 AM
191820  9/9/2019 9:10:00 PM
191832  9/9/2019 9:08:00 AM
192599  9/9/2019 9:06:00 AM
190713  9/9/2019 8:45:00 AM

<b>to 9/9/2019</b>

<h3>Removing unnecessary columns</h3>

<b>We're gonna focus on the location, the date and time, weather, road and light condition, collision type, if there were speeding, influence of drugs or alcohol and inattention.</b>

In [26]:
df = df[['SEVERITYCODE', 'ADDRTYPE', 'LOCATION', 'COLLISIONTYPE', 'INCDTTM', 'INATTENTIONIND', 'UNDERINFL', 'WEATHER',
        'ROADCOND', 'LIGHTCOND', 'SPEEDING']]
df.head()

SEVERITYCODE      ADDRTYPE                                     LOCATION  \
0             2  Intersection                   5TH AVE NE AND NE 103RD ST   
1             1         Block   AURORA BR BETWEEN RAYE ST AND BRIDGE WAY N   
2             1         Block  4TH AVE BETWEEN SENECA ST AND UNIVERSITY ST   
3             1         Block     2ND AVE BETWEEN MARION ST AND MADISON ST   
4             2  Intersection              SWIFT AVE S AND SWIFT AV OFF RP   

  COLLISIONTYPE                 INCDTTM INATTENTIONIND UNDERINFL   WEATHER  \
0        Angles    3/27/2013 2:54:00 PM            NaN         N  Overcast   
1     Sideswipe   12/20/2006 6:55:00 PM            NaN         0   Raining   
2    Parked Car  11/18/2004 10:20:00 AM            NaN         0  Overcast   
3         Other    3/29/2013 9:26:00 AM            NaN         N     Clear   
4        Angles    1/28/2004 8:04:00 AM            NaN         0   Raining   

  ROADCOND                LIGHTCOND SPEEDING  
0      Wet                 Daylight      NaN  
1      Wet  Dark - Street Lights On      NaN  
2      Dry                 Daylight      NaN  
3      Dry                 Daylight      NaN  
4      Wet                 Daylight      NaN

<h3>Data types</h3>

In [27]:
df.dtypes

SEVERITYCODE       int64
ADDRTYPE          object
LOCATION          object
COLLISIONTYPE     object
INCDTTM           object
INATTENTIONIND    object
UNDERINFL         object
WEATHER           object
ROADCOND          object
LIGHTCOND         object
SPEEDING          object
dtype: object

<h4>Cast to datetime</h4>

<b>The data type of INCDTTM needs to be cast to datetime.</b>

In [28]:
df = df.astype({'INCDTTM': 'datetime64'})
df[['INCDTTM']].dtypes

INCDTTM    datetime64[ns]
dtype: object

<h3>Descriptive analysis</h3>

In [32]:
df.describe()

SEVERITYCODE
count  194673.000000
mean        1.298901
std         0.457778
min         1.000000
25%         1.000000
50%         1.000000
75%         2.000000
max         2.000000

In [33]:
df.describe(include=['O'])

ADDRTYPE                                           LOCATION  \
count    192747                                             191996   
unique        3                                              24102   
top       Block  BATTERY ST TUNNEL NB BETWEEN ALASKAN WY VI NB ...   
freq     126926                                                276   

       COLLISIONTYPE INATTENTIONIND UNDERINFL WEATHER ROADCOND LIGHTCOND  \
count         189769          29805    189789  189592   189661    189503   
unique            10              1         4      11        9         9   
top       Parked Car              Y         N   Clear      Dry  Daylight   
freq           47987          29805    100274  111135   124510    116137   

       SPEEDING  
count      9333  
unique        1  
top           Y  
freq       9333

<b>There seems to be some missing values.</b>

<h3>Missing Values</h3>

In [34]:
df.isnull().sum()

SEVERITYCODE           0
ADDRTYPE            1926
LOCATION            2677
COLLISIONTYPE       4904
INCDTTM                0
INATTENTIONIND    164868
UNDERINFL           4884
WEATHER             5081
ROADCOND            5012
LIGHTCOND           5170
SPEEDING          185340
dtype: int64

<b>SPEEDING and INATTENTIONIND have a lot of missing values. SPEEDING indicates whether or not speeding was a factor in the
collision, and INATTENTIONIND indicates whether or not collision was due to inattention. Both column should display Y when the condition was present and N when it was not, but from the describe() method we can see that there's only one unique value for both of the columns and that value is Y. We can either assume that the missing values indicates N, or drop the columns. Let's try the first.</b>

In [43]:
df[['INATTENTIONIND','SPEEDING']] = df[['INATTENTIONIND','SPEEDING']].fillna('N')

In [44]:
df.isnull().sum()

SEVERITYCODE         0
ADDRTYPE          1926
LOCATION          2677
COLLISIONTYPE     4904
INCDTTM              0
INATTENTIONIND       0
UNDERINFL         4884
WEATHER           5081
ROADCOND          5012
LIGHTCOND         5170
SPEEDING             0
dtype: int64

In [45]:
df[['INATTENTIONIND','SPEEDING']].describe()

INATTENTIONIND SPEEDING
count          194673   194673
unique              2        2
top                 N        N
freq           164868   185340